lib/Yield/HexDump.cpp

```diff
@@ -0,0 +1,302 @@
+/// \file HexDump.cpp
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "boost/icl/interval_map.hpp"
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "revng/EarlyFunctionAnalysis/FunctionMetadataCache.h"
+#include "revng/Model/RawBinaryView.h"
+#include "revng/Pipeline/AllRegistries.h"
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipes/FileContainer.h"
+#include "revng/Pipes/Kinds.h"
+#include "revng/Pipes/Ranks.h"
+#include "revng/Support/IRHelpers.h"
+
+namespace boost::icl {
+
+template<>
+struct is_discrete<MetaAddress> {
+  using type = is_discrete;
+  BOOST_STATIC_CONSTANT(bool, value = true);
+};
+
+template<>
+struct identity_element<MetaAddress> {
+  static MetaAddress value() { return MetaAddress{}; }
+};
+
+template<>
+struct has_difference<MetaAddress> {
+  using type = has_difference;
+  BOOST_STATIC_CONSTANT(bool, value = true);
+};
+
+template<>
+struct difference_type_of<MetaAddress> {
+  using type = uint64_t; // assume we compare only .address() within same type,
+                         // epoch and address space
+};
+
+template<>
+struct size_type_of<MetaAddress> {
+  using type = uint64_t;
+};
+
+} // namespace boost::icl
+
+inline MetaAddress &operator++(MetaAddress &Address) {
+  return Address += 1u;
+}
+
+inline MetaAddress &operator--(MetaAddress &Address) {
+  return Address -= 1u;
+}
+
+namespace revng::pipes {
+
+static llvm::Error outputHexDump(const TupleTree<model::Binary> &Binary,
+                                 const pipeline::LLVMContainer &Module,
+                                 const BinaryFileContainer &SourceBinary,
+                                 llvm::StringRef OutputPath) {
+  auto BufferOrError = llvm::MemoryBuffer::getFileOrSTDIN(*SourceBinary.path());
+  auto Buffer = cantFail(errorOrToExpected(std::move(BufferOrError)));
+  RawBinaryView BinaryView(*Binary.get(), Buffer->getBuffer());
+
+  std::error_code ErrorCode;
+  llvm::raw_fd_ostream Output(OutputPath,
+                              ErrorCode,
+                              llvm::sys::fs::CD_CreateAlways);
+  if (!!ErrorCode) {
+    return llvm::make_error<llvm::StringError>("Could not open file "
+                                                 + OutputPath.str(),
+                                               ErrorCode);
+  }
```

Comment:
1. Drop `!!` if no necessary
2. Can you use `llvm::createStringError`?

Reply:
I took this `!!` from some other file in revng project :(